### PR TITLE
document set of array data wire format

### PR DESCRIPTION
**Pull request by fmoor**
fixes https://github.com/edgedb/edgedb/issues/2266

### CODE REVIEW

File: docs/internals/protocol/dataformats.rst [hunk 1]
```diff
@@ -53,6 +53,40 @@ Note: zero-length arrays (and sets) are represented as a 12-byte value where
...
 
+Sets of arrays are a special case and are represented as the following
+structure:
+
+.. code-block:: c
+
+    struct SetOfArrayValue {
+        // Number of dimensions, currently must
+        // always be 0 or 1.
+        int32 ndims;
```
Comment (1st1) [inline docs/internals/protocol/dataformats.rst:64]: Maybe add newlines after every field?

File: docs/internals/protocol/dataformats.rst [hunk 2]
```diff
@@ -53,6 +53,40 @@ Note: zero-length arrays (and sets) are represented as a 12-byte value where
...
 
 
+Sets of arrays are a special case and are represented as the following
+structure:
+
+.. code-block:: c
+
+    struct SetOfArrayValue {
+        // Number of dimensions, currently must
+        // always be 0 or 1.
```
Comment (1st1) [inline docs/internals/protocol/dataformats.rst:63]: Maybe clarify what dimensions 0 and 1 mean?

**Review by 1st1 — APPROVED** [review]

File: docs/internals/protocol/dataformats.rst (hunk 2 above)
Comment (fmoor) [reply docs/internals/protocol/dataformats.rst:63]: I'm not sure what I would say besides `0 means 0 dimensions and 1 means 1 dimension.` Maybe I don't fully understand their significance.

File: docs/internals/protocol/dataformats.rst (hunk 1 above)
Comment (fmoor) [reply docs/internals/protocol/dataformats.rst:70]: changed

File: docs/internals/protocol/dataformats.rst [hunk 3]
```diff
@@ -53,6 +59,47 @@ Note: zero-length arrays (and sets) are represented as a 12-byte value where
 ``dims`` equal to zero regardless of the shape in type descriptor.
 
 
+Sets of arrays are a special case and are represented as the following
```
Comment (tailhook) [inline docs/internals/protocol/dataformats.rst:62]: I'm not sure it's easy to find out what is `Envelope` and what is `Element` and what is the difference. So some intro or comment should state something like "every element of the set of arrays is wrapped into an envelope".

It's easier to think about it like "every array within a set is wrapped into a tuple of size one", but I don't think it fits the narrative here.

Comment (fmoor) [reply docs/internals/protocol/dataformats.rst]: changed

File: docs/internals/protocol/dataformats.rst (hunk 2 above)
Comment (fmoor) [reply docs/internals/protocol/dataformats.rst:69]: @1st1 is there something specific you were thinking here?

Comment (elprans) [reply docs/internals/protocol/dataformats.rst:69]: I'd clarify that zero dimensions is just an empty array.

Comment (1st1) [reply docs/internals/protocol/dataformats.rst:69]: What @elprans said :)

Comment (fmoor) [reply docs/internals/protocol/dataformats.rst]: changed

**Pull request merged**